src/snmp.c

```diff
@@ -1388,6 +1388,50 @@ static int csnmp_read_table(host_definition_t *host, data_definition_t *data) {
       break;
     }
 
+    if (res->errstat != SNMP_ERR_NOERROR) {
+      if (res->errindex != 0) {
+        /* Find the OID which caused error */
+        for (i = 1, vb = res->variables; vb != NULL && i != res->errindex;
+             vb = vb->next_variable, i++)
+          /* do nothing */;
+      }
+
+      if ((res->errindex == 0) || (vb == NULL)) {
+        ERROR("snmp plugin: host %s; data %s: response error: %s (%li) ",
+              host->name, data->name, snmp_errstring(res->errstat),
+              res->errstat);
+        status = -1;
+        break;
+      }
+
+      char oid_buffer[1024] = {0};
+      snprint_objid(oid_buffer, sizeof(oid_buffer) - 1, vb->name,
+                    vb->name_length);
+      NOTICE("snmp plugin: host %s; data %s: OID `%s` failed: %s", host->name,
+             data->name, oid_buffer, snmp_errstring(res->errstat));
+
+      /* Calculate value index from todo list and skip OID found */
+      i = 0;
```

Comment:
One way to simplify this might be with a temporary array:

```c
size_t oid_active[oid_list_len] = {0};
size_t oid_active_len = 0;

for (size_t i = 0; i < oid_list_len; i++) {
  if (!oid_list_todo[i]) {
    continue;
  }
  oid_active[oid_active_len] = i;
  oid_active_len++;
}

assert(res->errindex <= oid_active_len);
size_t failed_index = oid_active[res->errindex - 1];
oid_list_todo[failed_index] = 0;
``` 

It's also quite long, but IMHO easier to read than two indexes that get incremented in two nested loops with conditionals …

Reply:
So, the same `oid_active` should be used in next cycle (https://github.com/collectd/collectd/pull/2449/files#diff-15314dd4787306bec3b25576017b94b3R1435) too? That will do things better, IMHO.

Reply:
Good eye! Yes, that sounds like an improvement. (There are many improvements to be made to the *snmp plugin's* code ;)

Reply:
Hmm, if use such array for 'next cycle' then 'two indexes' will appear there.
I prefer to leave 'next cycle' untouched.

Don't you think it is superfluous to form a temporary array for rare errors?
I think that this is not optimal. May be left that cycle untouched, without temporary array  but with two indexes?